File: _1448_count_good_nodes_in_binary_tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def visit(self, root: TreeNode, max: int) -> int:
        if not root:
            return 0

        count = 0

        if root.val >= max:
            count = 1
            max = root.val

        count += self.visit(root.left, max)
        count += self.visit(root.right, max)

        return count

    def goodNodes(self, root: TreeNode) -> int:
        return self.visit(root, -math.inf)

File: _1448_count_good_nodes_in_binary_tree/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
        return explore(root, INT_MIN);
    }

    int explore(TreeNode* root, int maximum) {
        if (root == NULL) {
            return 0;
        }

        int count = 0;
        if (root->val >= maximum) {
            count = 1;
        }

        return count + explore(root->left, max(root->val, maximum)) +  explore(root->right, max(root->val, maximum));
    }
};